src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,597 @@ uint AddPNode::match_edge(uint idx) const {
+  MemoryAdjacentStatus get_adjacent_load_status(const LoadNode* first, const LoadNode* second) const;
+  // Go through ConvI2L which is unique output of input node
+  static const Node* bypass_i2l(const Node* n);
+  // Add operators of combine operator to IGVN worklist
+  void add_operators_to_worklist(const AddNode* combine) const;
+
+#ifndef PRODUCT
+  // Access to TraceMergeLoads tags
+  bool is_trace(TraceMergeLoads::Tag tag) const {
+    return _trace_tags.at(tag);
+  }
+
+  bool is_trace_basic() const {
+    return is_trace(TraceMergeLoads::Tag::BASIC);
+  }
+
+  bool is_trace_pointer_parsing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_PARSING);
+  }
+
+  bool is_trace_pointer_aliasing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ALIASING);
+  }
+
+  bool is_trace_pointer_adjacency() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ADJACENCY);
+  }
+
+  bool is_trace_success() const {
+    return is_trace(TraceMergeLoads::Tag::SUCCESS);
+  }
+#endif
+};
+
+bool MergePrimitiveLoads::is_supported_load_opcode(int opc) {
+  // Check for B/S/C/I
+  switch(opc) {
+    case Op_LoadB:
+    case Op_LoadUB:
+    case Op_LoadS:
+    case Op_LoadUS:
+    case Op_LoadI:
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool MergePrimitiveLoads::is_supported_combine_opcode(int opc) {
+  switch(opc) {
+    case Op_OrI:
+    case Op_OrL:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// Go through ConvI2L which is unique output of input node
+const Node* MergePrimitiveLoads::bypass_i2l(const Node* n) {
+  if (n != nullptr && n->outcnt() == 1 && n->unique_out()->Opcode() == Op_ConvI2L) {
+    return n->unique_out();
+  } else {
+    return n;
+  }
+};
+
+/*
+ * Check the _combine operator is the last one of combine operators
+ * It can be a candidate for merge load optimization
+ */
+bool MergePrimitiveLoads::has_no_merge_load_combine_below() const {
+  assert(is_supported_combine_opcode(_combine->Opcode()), "sanity");
+  const Node* check = bypass_i2l(_combine);
+  if (check->outcnt() == 1 &&
+      check->unique_out()->Opcode() == _combine->Opcode()) {
+    AddNode* out = check->unique_out()->as_Add();
+    if (out->is_merge_memops_checked()) {
+      // the next operator is checked before, so _combine can be last one of combine operators
+      return true;
+    } else {
+      // It's in the middle of combine operators
+      return false;
+    }
+  }
+  return true;
+}
+
+// Construct merge information item from input load
+MergeLoadInfo MergePrimitiveLoads::merge_load_info(LoadNode* load) const {
+  const MergeLoadInfo invalid = MergeLoadInfo();
+  const Node* check = bypass_i2l(load);
+  Node* combine_oper = nullptr;
+  jint shift = -1;
+
+  // Check the Load node has the pattern "(Or (LShift (Load .. ) ConI) ..)" or "(Or (Load ..) ..)"
+  for (DUIterator_Fast imax, iter = check->fast_outs(imax); iter < imax; iter++) {
+    Node *out = check->fast_out(iter);
```

Comment:
Do you need to loop here?
Do all cases we expect to optimize only have a single use of `check`?
Or what exactly would happen here if we had multiple uses?

Could be good if you had a regression test that triggers such a case with multiple uses here.

Reply:
Yeah, it seems you actually check when looping if you find a second one, and then return `invalid`. And the default case also gets `invalid`. Hence, I really see no point in looping here, you should just use `unique_out`, right?